### ISSUE TITLE
C++ API for Transformer model in libtorch 1.5.0

### ISSUE THREAD

**Issue by srnvsk** [pytorch/pytorch#37756]
In the latest 1.5 release, I am unable to find Transformer model api in C++ (equivalent to the torch.nn.Transformer in its python counterpart). Any timeline for its implementation in C++?

cc @yf225 @glaringlee

**Comment by dmarx** [comment]
Looks like that addition is pending a refactor to the python implementation. See here for tracking: https://github.com/pytorch/pytorch/issues/25883

**Comment by glaringlee** [comment]
@srnvsk
Transformer model python end is refactoring, we want to implement c++ version after the refactor so we don't need to modify it later. Will check the progress of the refactoring and get back here later.

**Comment by srnvsk** [comment]
Thanks for your reply!

**Comment by VinodSKumar** [comment]
> @srnvsk
> Transformer model python end is refactoring, we want to implement c++ version after the refactor so we don't need to modify it later. Will check the progress of the refactoring and get back here later.

@glaringlee could you please explain what refactoring needs to be done. I could contribute to it. Then the Python/C++ API parity issue [25883](https://github.com/pytorch/pytorch/issues/25883) could be addressed.

**Comment by glaringlee** [comment]
@VinodSKumar
I should update this thread, there is no longer refactoring now, the c++ transformer will be added into C++ Frontend within the rest of the year, I am currently working on this. Let me know if you want to contribute in adding pieces of transformer layer to C++, the work is definitely parallel-able.

Some background on refactoring. When transformer module is first added to pytorch in python, we were planing to add it to c++ API as well, then we stopped since there was a discussion on whether python version transformer need a refactor. And then we have an internal position shift so the discussion stopped for a while. I am working on c++ api now and re-discussed with the owner of transformer python version owner, we decided not to refactor it. I will add the entire transformer module to c++ within the rest of the year.

**Comment by VinodSKumar** [comment]
@glaringlee I could work on adding the pieces of transformer layer to C++, please let me know the parts that I could focus on.

**Comment by glaringlee** [comment]
@VinodSKumar 
I am currently working on transformerEncoderLayer and transformerEncoder. You can focus on TransformerDecoderLayer, and then if you have time and want to continue, you can continue working on TransformerDecoder. There will be no infrastructure change, we already have all the needed models (linear, dropout, normlayer). You can take this as an example.
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/modules/adaptive.h#L41
I will have a draft PR for transformerEncoderLayer later this week or earlier next week, you can also take that as a reference.

**Comment by VinodSKumar** [comment]
Thank you will look into it.

**Comment by VinodSKumar** [comment]
@glaringlee could you please help to confirm this issue [42424](https://github.com/pytorch/pytorch/issues/42424). For TransformerDecoderLayer UT, we need to initialize the (sub)modules parameters.

**Comment by VinodSKumar** [comment]
@glaringlee Request your help to review the PR [42587](https://github.com/pytorch/pytorch/pull/42587) for TransformerDecoderLayer 

In the meantime, I will look into the implementation of  TransformerDecoder

**Comment by glaringlee** [comment]
@VinodSKumar 
I will take a look at it today. Thanks for help!

**Comment by VinodSKumar** [comment]
@glaringlee There was some issue with the old PR in git, I have created a new PR  [42717](https://github.com/pytorch/pytorch/pull/42717) addressing all the review comments from PR [42587](https://github.com/pytorch/pytorch/pull/42587) sorry for the trouble of raising a new PR, could you please help to continue the review on the new PR.

**Comment by glaringlee** [comment]
@VinodSKumar I see your previous PR had some rebase issue and you close it.
No problem, will take a look at tomorrow.

**Comment by glaringlee** [comment]
Just FYI. This is available in the master branch already. #44333 
We are working on turning on the parity test at the moment
